### PR TITLE
Weather repository

### PR DESCRIPTION
**Pull request by deffsin**

### CODE REVIEW

**Review by Hassanq911 — CHANGES_REQUESTED** [review]

File: WeatherApp/Services/CoreData/WeatherRepository.swift [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    func saveWeatherData(_ weatherData: WeatherDto) {
+        if let existingWeather = coreDataService.find(Weather.self, with: weatherData.city, forKey: "city") {
+            existingWeather.city = weatherData.city
+            existingWeather.country = weatherData.country
+            existingWeather.weatherDescription = weatherData.weatherDescription
+            existingWeather.temperature = weatherData.temperature
+            existingWeather.sunrise = weatherData.sunrise
+            existingWeather.sunset = weatherData.sunset
+        } else {
+            let weatherEntity = Weather(context: context)
```
[the remark this answers is not in the document]
Comment (deffsin) [reply WeatherApp/Services/CoreData/WeatherRepository.swift:30]: What's the reason to use extension for that? Make code much cleaner and readable?

[the remark this answers is not in the document]
Comment (Hassanq911) [reply WeatherApp/Services/CoreData/WeatherRepository.swift:25]: I would say redundancy so in future if we have lets say more entities we wont need to write out the mapping manually

File: WeatherApp/Services/CoreData/WeatherRepository.swift [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    private let coreDataService: CoreDataService
+    private let context: NSManagedObjectContext
+    
+    init(coreDataService: CoreDataService) {
+        self.coreDataService = coreDataService
+        self.context = coreDataService.context
+    }
+    
+    func saveWeatherData(_ weatherData: WeatherDto) {
+        if let existingWeather = coreDataService.find(Weather.self, with: weatherData.city, forKey: "city") {
```
[the remark this answers is not in the document]
Comment (Hassanq911) [reply WeatherApp/Services/CoreData/WeatherRepository.swift:22]:
```
let weatherEntity = coreDataService.find(Weather.self, with: weatherData.city, forKey: "city") ?? Weather(context: context)
weatherData.mapToWeatherEntity(weatherEntity)
```
maybe something like this to make it concise.

**Review by Hassanq911 — CHANGES_REQUESTED** [review]

File: WeatherApp/Services/CoreData/WeatherRepository.swift [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    private let coreDataService: CoreDataService
+    private let context: NSManagedObjectContext
+    
+    init(coreDataService: CoreDataService) {
+        self.coreDataService = coreDataService
+        self.context = coreDataService.context
+    }
+    
+    func saveWeatherData(_ weatherData: WeatherDto) {
+        let weatherEntity = coreDataService.find(Weather.self, with: weatherData.city, forKey: "city") ?? Weather(context: context)
```
Comment (Hassanq911) [inline WeatherApp/Services/CoreData/WeatherRepository.swift:22]: Referring to my previous comment find a way that we don't have to explicitly write the attribute name for which we want to search the entity for rather just pass in the entity object in the coredata service's function and it does the rest. 
I was thinking for implementing a protocol for this purpose if their are no native solutions

File: WeatherApp/Services/CoreData/WeatherRepository.swift [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    
+    func getWeatherForCity(_ city: String) -> WeatherDto? {
+            guard let weatherEntity = coreDataService.find(Weather.self, with: city, forKey: "city") else { return nil }
+            return weatherEntity.toWeatherDto()
+        }
+        
+        func getAllWeatherData() -> [WeatherDto] {
+            let weatherEntities = coreDataService.fetchAll(Weather.self)
```
Comment (Hassanq911) [inline WeatherApp/Services/CoreData/WeatherRepository.swift:33]: This function can be a 1 liner 
`coreDataService.fetchAll(Weather.self).map { $0.toWeatherDto() }`

File: WeatherApp/Services/CoreData/WeatherRepository.swift [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    }
+    
+    func saveWeatherData(_ weatherData: WeatherDto) {
+        let weatherEntity = coreDataService.find(Weather.self, with: weatherData.city, forKey: "city") ?? Weather(context: context)
+        weatherData.mapToWeatherEntity(weatherEntity)
+    }
+
+    
+    func getWeatherForCity(_ city: String) -> WeatherDto? {
+            guard let weatherEntity = coreDataService.find(Weather.self, with: city, forKey: "city") else { return nil }
```
Comment (Hassanq911) [inline WeatherApp/Services/CoreData/WeatherRepository.swift:28]: This function can be a 1 liner
`coreDataService.find(Weather.self, with: city, forKey: "city")?.toWeatherDto()`

File: WeatherApp/Services/CoreData/WeatherRepository.swift (hunk 3 above)
Comment (deffsin) [reply WeatherApp/Services/CoreData/WeatherRepository.swift:22]: Well, i couldn't find another way besides using a protocol, so i'll try to using it

**Review by Hassanq911 — APPROVED** [review]

**Pull request closed without merging**